GUACAMOLE-919: Implement PostgreSQL defaultStatementTimeout and socketTimeout

DouglasHeriot opened the pull request:
Pass through environment variables to set `defaultStatementTimeout` and `socketTimeout`. See issue GUACAMOLE-919 for details.

If the TCP connection to the database is silently dropped (not closed), doing a ping query "SELECT 1;" will hang indefinitely. The socketTimeout option will put a timeout on this.

`defaultStatementTimeout` is not relevant to this issue. I thought it might have been at first. I've left this in incase having the option exposed is useful for other reasons.

necouchman commented on extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-postgresql/src/main/java/org/apache/guacamole/auth/postgresql/PostgreSQLEnvironment.java:
```diff
@@ -47,6 +47,21 @@
+     * https://jdbc.postgresql.org/documentation/head/connect.html
+     */
+    private static final int DEFAULT_SOCKET_TIMEOUT = 0;
+
```
Is there some reason why you've implemented one of these as a `String` and the other as an `int`?  I think I would stick with `int` to be consistent with how other timeouts are implemented.  Also, one uses `null` for no timeout, the other uses 0 - can we align on that, or does the underlying code note allow for that?

DouglasHeriot replied:
Yes, I was trying to keep same as what the backend takes. But I see your point - consistency at this level would be good rather than exposing the underlying differences.

necouchman commented on extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-postgresql/src/main/java/org/apache/guacamole/auth/postgresql/PostgreSQLEnvironment.java:
```diff
@@ -47,6 +47,21 @@
+     * Default to null (no timeout)
+     * https://mybatis.org/mybatis-3/configuration.html
+     */
+    private static final String DEFAULT_DEFAULT_STATEMENT_TIMEOUT = "null";
```
`"null"` != `NULL` - does the underlying setting require the String value `null` for this?

DouglasHeriot replied:
Yes it takes a string value. I thought that just keeping it as a string was neater than making a nullable `Integer` where all other properties are `int`.
I've now changed this to an `int` and don't pass through when it's `0`.

necouchman replied:
I dug into this a little more, and it does not look to me like the string value "null" is what they intend - I think just mean `NULL` in the since of not setting anything.  So, this should not be `"null"`, it should be `NULL`.

DouglasHeriot replied:
@necouchman I agree, it looks like I made a mistake. See my latest commit on this branch for the fix 330b2c3ec0c1fa5ed62f4efcb8585f7c287f86c9

DouglasHeriot commented:
These properties are passed in as string properties. `"null"` means not-set for `defaultStatementTimeout` however `0` is not-set for `socketTimeout`. I initially decided to try and just keep the properties as strings, but `int` did make more sense for `socketTimeout`.

https://mybatis.org/mybatis-3/configuration.html
https://jdbc.postgresql.org/documentation/head/connect.html

I agree int is nicer, so to keep consistency I have now made `defaultStatementTimeout` and `int` as well, and simply don't set the property on the backend if it's `0` to match the behaviour of `socketTimeout`.

necouchman commented on extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-postgresql/src/main/java/org/apache/guacamole/auth/postgresql/PostgreSQLGuacamoleProperties.java:
```diff
@@ -92,6 +92,30 @@ private PostgreSQLGuacamoleProperties() {}
     };
 
+    /**
+     * Sets the number of seconds the driver will wait for
```
This is a bit of a nitpick, but the below statement doesn't actually "set" anything.  It is a property that can be used to set it.  Take a look at some of the other properties around this - for example, "The password used to authenticate..."  This can probably be simplified to something similar to that.

necouchman commented on extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-postgresql/src/main/java/org/apache/guacamole/auth/postgresql/PostgreSQLGuacamoleProperties.java:
```diff
@@ -92,6 +92,30 @@ private PostgreSQLGuacamoleProperties() {}
+    };
+
+    /**
+     * Sets the number of seconds the driver will wait in a read() call
```
As above, this isn't setting anything, so more direct language would probably be better, here.

necouchman commented on extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-postgresql/src/main/java/org/apache/guacamole/auth/postgresql/PostgreSQLEnvironment.java:
```diff
@@ -47,6 +47,21 @@
+     * The default defaultStatementTimeout (in seconds),
+     * if POSTGRESQL_DEFAULT_STATEMENT_TIMEOUT is not specified.
+     * Default to 0 (no timeout, property won't be set)
+     * https://mybatis.org/mybatis-3/configuration.html
```
A couple things for this comment block:
* I'd avoid using the actual variable name (`defaultStatementTimeout`) and just make it read like "The default statement timeout..."
* We don't really link to outside documentation for any of the other properties, and I'm not sure it's worth starting for these two, so I'd remove the links to external sites.
* For the explanation for "Default to 0," something more verbose might be appropriate: "Setting this value to 0 (the default) will result in..."

DouglasHeriot commented:
Ok, sorry @necouchman for the delay.
I've fixed the style issues and improved the comments.

I've also added a corresponding pull request to the manual repo:
https://github.com/apache/guacamole-manual/pull/146

necouchman commented on extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-postgresql/src/main/java/org/apache/guacamole/auth/postgresql/conf/PostgreSQLGuacamoleProperties.java:
```diff
@@ -94,6 +94,30 @@ private PostgreSQLGuacamoleProperties() {}
+     * a response from the database.
+     */
+    public static final IntegerGuacamoleProperty
+            POSTGRESQL_DEFAULT_STATEMENT_TIMEOUT = new IntegerGuacamoleProperty(){
```
Please put a space between `()` and `{`.

necouchman commented on extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-postgresql/src/main/java/org/apache/guacamole/auth/postgresql/conf/PostgreSQLGuacamoleProperties.java:
```diff
@@ -94,6 +94,30 @@ private PostgreSQLGuacamoleProperties() {}
+     * on the TCP connection to the database.
+     */
+    public static final IntegerGuacamoleProperty
+            POSTGRESQL_SOCKET_TIMEOUT = new IntegerGuacamoleProperty(){
```
Space between `()` and `{`, please.

necouchman commented on guacamole-docker/bin/start.sh:
```diff
@@ -354,10 +354,18 @@ END
         "postgresql-default-max-group-connections-per-user" \
         "$POSTGRES_DEFAULT_MAX_GROUP_CONNECTIONS_PER_USER"
 
+    set_optional_property               \
+        "postgresql-default-statement-timeout" \
```
Please line up the trailing `\`.

necouchman replied:
These two lines still do not appear to line up the trailing slashes.

necouchman commented on guacamole-docker/bin/start.sh:
```diff
@@ -354,10 +354,18 @@ END
         "postgresql-user-required" \
         "$POSTGRES_USER_REQUIRED"
 
+    set_optional_property               \
+        "postgresql-socket-timeout" \
```
Please line up the trailing `\`

necouchman replied:
Trailing slashes are still not lined up, here.

necouchman commented on extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-postgresql/src/main/java/org/apache/guacamole/auth/postgresql/conf/PostgreSQLEnvironment.java:
```diff
@@ -249,6 +265,41 @@ public String getPostgreSQLUsername() throws GuacamoleException {
+    
+    /**
+     * Returns the socketTimeout property to set on PostgreSQL connections.
+     * If unspecified, this will be the default to 0 (no timeout)
```
Probably should either be:
> this will default to

or:
> this will be the default of

necouchman commented on extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-postgresql/src/main/java/org/apache/guacamole/auth/postgresql/conf/PostgreSQLEnvironment.java:
```diff
@@ -48,6 +48,22 @@
+     * If reading from the server takes longer than this value, the
+     * connection is closed. This can be used to handle network problems
+     * such as a dropped connection to the database. Similar to 
+     * DEFAULT_STATEMENT_TIMEOUT, it will also abort queries that take too 
```
Two comments, here:
* You say "similar to DEFAULT_STATEMENT_TIMEOUT", but there's no mention of aborting queries on the above timeout.
* It looks like most of the documentation around what these two additional parameters do is here in the `DEFAULT_` sections, rather than down in the `PostgreSQLGuacamoleProperties.java` file.  If you look at the documentation of other `DEFAULT_`s around this, you'll see that the documentation is more succinct and references the actual properties.  I suggest doing the same with these two.

necouchman commented on extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-postgresql/src/main/java/org/apache/guacamole/auth/postgresql/PostgreSQLAuthenticationProviderModule.java:
```diff
@@ -70,6 +70,12 @@ public PostgreSQLAuthenticationProviderModule(PostgreSQLEnvironment environment)
+        // Only set if > 0. Underlying backend does not take 0 as not-set.
+        int defaultStatementTimeout = environment.getPostgreSQLDefaultStatementTimeout();
+        if (defaultStatementTimeout > 0) {
+            myBatisProperties.setProperty("mybatis.configuration.defaultStatementTimeout", String.valueOf(defaultStatementTimeout));
```
Might be good to break this line somewhere - this is a looooooooooooooong line :-).

DouglasHeriot commented:
Ok @necouchman , I've addressed the style issues, and moved most of the detail in the comments to `PostgreSQLGuacamoleProperties.java`.

I've also updated apache/guacamole-manual#146 to match the new wording.

DouglasHeriot commented:
@necouchman Oops, sorry. I made the change and didn't add it to my earlier commit. Fixed now.

necouchman commented:
Okay, @DouglasHeriot, the only remaining issue I see is that your commit messages are not properly formatted.  The style of the commit message should be:
> GUACAMOLE-919: Description of the changes in the commit

The ":" appears to be missing from your commit messages.  Once that's fixed we should be ready to merge.

DouglasHeriot commented:
@necouchman  Ok, I've fixed these commit messages now.
